rfc/cart-api-v1.md

```diff
@@ -0,0 +1,179 @@
+| --- | --- | --- |
+| `POST` | `/resources/carts/` | Create a new (non-default) cart |
+| `GET` | `/resources/carts/` | Retrieve a list of carts |
+| `GET` | `/resources/carts/{id}` | Retrieve a single cart (see more detail below) |
+| `PUT` | `/resources/carts/{id}` | Update the cart (see more detail below) |
+| `DELETE` | `/resources/carts/{id}` | Delete a single cart (see more detail below) |
+
+#### `GET /resources/carts/{id}`
+
+Retrieve a single cart.
+
+##### Response
+
+Instead of returning `Cart` information, the response will be also returning the list of associated `CartItem`, as shown below:
+
+```javascript
+{
+    id: str, // Cart ID
+    cart_name: str,
+    default: bool,
+    items: List[CartItem] // Each CartItem object will not have `cart_id`.
+}
+```
+
+##### Default Cart
+
+The default cart is a special case and will be given back to the client ONLY IF `id` is `default` (all lowercase).
+
+* `default` is **the alias ID** of the default cart belonging to the session user.
+  * The actual ID of default cart will be returned in the response body.
+  * This ID is only used in this endpoint.
+* When the default cart is not available, the default cart will be created and return in the same request.
+
+Here is the pseudo code for the request handler.
+
+```python
+# Pseudo code: Request Handler
+def retrieve_one_cart(id: UUID):
+    # Define: bearer_token is the request bearer token
+    if id == 'default':
+        cart = cart_repository.get_default(user_id = bearer_token)
+
+        if not cart:
+            # Either this or: cart = cart_repository.create(user_id = bearer_token, cart_name = '')
+            cart = cart_repository.create(user_id = bearer_token, cart_name = '', default = True)
+    else:
+        cart = cart_repository.get(id)
+
+    return dict(id=cart.id,
+                cart_name=cart.cart_name,
+                entity_type=cart.entity_type,
+                default=cart.default,
+                items=cart_item_repository.find_many_by_cart_id(cart.id))
+```
+
+Here is the sample request.
+
+```
+GET /resources/carts/default
+Authorization: Bearer 261e1aea-82c9-4e38-a879-e2e02ef236a5
+```
+
+Here is the sample response.
+
+```javascript
+{
+    id: "6b801de6-11d4-46c0-aed1-ca2110aa7d3c"
+    cart_name: "",
+    default: true,
+    items: [...]
+}
+```
+
+#### `PUT /resources/carts/{id}`
+
+Here is the request body.
+
+```javascript
+{
+    cart_name: str // Required
+}
+```
+
+and the response is basically the JSON representation of `Cart`.
+
+#### `DELETE /resources/carts/{id}`
+
+* The endpoint will only delete the target cart if the user creates the target cart and responds with **HTTP 200**.
+  * Otherwise, the endpoint must respond with **HTTP 403**.
+* The endpoint will respond **HTTP 410** (gone) if the cart is unknown.
+
+### CRUD APIs for `CartItem`
+
+| Method | Path | Description |
+| --- | --- | --- |
+| `POST` | `/resources/carts/{cart_id}/items/` | Create a new cart item |
```

Comment:
It would be useful to allow creation of multiple items in one request for adding results of a search.
We could use take a request body that has an items property that is an array of item objects.  To add a single item it would be an array with one element to stay consistent.

Reply:
We will have a batch API for that if the UI can add items in batch.

Reply:
Writing down our earlier discussion: In addition to a batch api we may have to take a query in the response, make the query in the backend, and then save the results to the cart.  This is to deal with pagination because if the user wants to "save all results to cart" the client may not have all the results loaded.